src/lib/util/ioprocs.cpp

```diff
@@ -0,0 +1,1232 @@
+	return result;
+}
+
+random_read::ptr stdio_read_noclose(FILE *file, std::uint8_t filler) noexcept
+{
+	std::unique_ptr<random_read_fill_wrapper<stdio_read_adapter> > result;
+	if (file)
+		result.reset(new (std::nothrow) decltype(result)::element_type(file, false));
+	if (result)
+		result->set_filler(filler);
+	return result;
+}
+
+
+// creating stdio read/write adapters
+
+random_read_write::ptr stdio_read_write(FILE *file) noexcept
+{
+	random_read_write::ptr result;
+	if (file)
+		result.reset(new (std::nothrow) stdio_read_write_adapter(file, true));
+	return result;
+}
+
+random_read_write::ptr stdio_read_write(FILE *file, std::uint8_t filler) noexcept
+{
+	random_read_write::ptr result;
+	if (file)
+		result.reset(new (std::nothrow) stdio_read_write_filler(file, true, filler));
+	return result;
+}
+
+random_read_write::ptr stdio_read_write_noclose(FILE *file) noexcept
+{
+	random_read_write::ptr result;
+	if (file)
+		result.reset(new (std::nothrow) stdio_read_write_adapter(file, false));
+	return result;
+}
+
+random_read_write::ptr stdio_read_write_noclose(FILE *file, std::uint8_t filler) noexcept
+{
+	random_read_write::ptr result;
+	if (file)
+		result.reset(new (std::nothrow) stdio_read_write_filler(file, false, filler));
+	return result;
+}
+
+
+// creating osd_file read adapters
+
+random_read::ptr osd_file_read(osd_file::ptr &&file) noexcept
+{
+	random_read::ptr result;
+	if (file)
+		result.reset(new (std::nothrow) osd_file_read_adapter(std::move(file)));
+	return result;
+}
+
+random_read::ptr osd_file_read(osd_file &file) noexcept
+{
+	return random_read::ptr(new (std::nothrow) osd_file_read_adapter(file));
+}
+
+
+// creating osd_file read/write adapters
+
+random_read_write::ptr osd_file_read_write(osd_file::ptr &&file) noexcept
+{
+	random_read_write::ptr result;
+	if (file)
+		result.reset(new (std::nothrow) osd_file_read_write_adapter(std::move(file)));
+	return result;
+}
+
+random_read_write::ptr osd_file_read_write(osd_file &file) noexcept
+{
+	return random_read_write::ptr(new (std::nothrow) osd_file_read_write_adapter(file));
+}
+
+
+// creating core_file read adapters
+
+random_read::ptr core_file_read(core_file::ptr &&file) noexcept
+{
+	random_read::ptr result;
+	if (file)
+		result.reset(new (std::nothrow) core_file_read_adapter(std::move(file)));
+	return result;
+}
+
+random_read::ptr core_file_read(core_file::ptr &&file, std::uint8_t filler) noexcept
+{
+	std::unique_ptr<random_read_fill_wrapper<core_file_read_adapter> > result;
+	if (file)
+		result.reset(new (std::nothrow) decltype(result)::element_type(std::move(file)));
+	if (result)
+		result->set_filler(filler);
+	return result;
+}
+
+random_read::ptr core_file_read(core_file &file) noexcept
+{
+	return random_read::ptr(new (std::nothrow) core_file_read_adapter(file));
+}
+
+random_read::ptr core_file_read(core_file &file, std::uint8_t filler) noexcept
+{
+	std::unique_ptr<random_read_fill_wrapper<core_file_read_adapter> > result(
```

Comment:
MSVC doesn't like using decltype() within a combined declaration/definition; logged [here](https://developercommunity.visualstudio.com/t/C2065:-undeclared-identifier-using-declt/1506615).

Reply:
I split it in two anyway.  It doesn’t really hurt readability, and everything’s `noexcept` there so it should optimise to more-or-less the same thing.